### ISSUE TITLE
gc might put vector tag on (intermediate?) procedure pointer

### ISSUE THREAD

**Issue by larceny-trac-import** [larcenists/larceny#338]
**Reported by: lth  on Fri Oct 15 06:00:00 1999**
138 v0.45 (991015 / lth)
Priority: low
Category: RTS / correctness

Description.

I have discovered a problem in the GC write barrier in the NP collector.
In the macro scan_core_partial() a pointer is added to the SSB of the NP
extra set by another macro remember_vec(), the call to that macro is:

(line 396)
   if (must_add_to_extra) remember_vec( tagptr( T_objp, VEC_TAG ), e );

However the case being handled in scan_core_partial() is that for both
procedures and vectors, thus we see that the pointer added to the SSB
and remembered set may have the wrong tag (a vector tag for a procedure
pointer).

Impact.

The problem only occurs when a procedure structure promoted to the GC
young area contains a pointer to the GC old area: in this case only will
an incorrect pointer be added to the SSB for the NP extra set.  Normally
procedures are created on-the-fly by loading compiled code, by creating
closures, by the interpreter (which actually creates them at a very high
rate, which is why some scrupulous debugging code found the problem),
and by the debugger.

The GC currently treats vectors and procedures alike by stripping the
tag and then handling the raw pointer, so I do not expect GC crashes due
to mis-tagging.

While the remembered set hashes the pointer by stripping the tag,
comparisons in the hash table are on the full pointer, so there could
plausibly be duplicate entries in the hash table.  Duplicates are
unlikely however since promotion creates only one pointer and the other
would have to be created by assignment, and procedures are usually
side-effected only for initialization and those assignments should
almost never trap in the write barrier.  If duplicates do occur, then an
object may be scanned twice.  As far as I know, that does not create any
problems for the collector.

In summary it is extremely unlikely that the bug causes performance
degradation, and as far as I can tell, it is not possible for it to
cause GC crashes.

Action.

The bug should be fixed.  Fixing it will slow down the case when a
pointer needs to be added to the SSB, probably not an issue in practice.
